src/openssl/ssl/socket.cr

```diff
@@ -2,36 +2,40 @@ abstract class OpenSSL::SSL::Socket < IO
   class Client < Socket
     def initialize(io, context : Context::Client = Context::Client.new, sync_close : Bool = false, hostname : String? = nil)
       super(io, context, sync_close)
-
-      if hostname
-        # Macro from OpenSSL: SSL_ctrl(s,SSL_CTRL_SET_TLSEXT_HOSTNAME,TLSEXT_NAMETYPE_host_name,(char *)name)
-        LibSSL.ssl_ctrl(
-          @ssl,
-          LibSSL::SSLCtrl::SET_TLSEXT_HOSTNAME,
-          LibSSL::TLSExt::NAMETYPE_host_name,
-          hostname.to_unsafe.as(Pointer(Void))
-        )
-
-        {% if compare_versions(LibSSL::OPENSSL_VERSION, "1.0.2") >= 0 %}
-          param = LibSSL.ssl_get0_param(@ssl)
-
-          if ::Socket.ip?(hostname)
-            unless LibCrypto.x509_verify_param_set1_ip_asc(param, hostname) == 1
-              raise OpenSSL::Error.new("X509_VERIFY_PARAM_set1_ip_asc")
-            end
-          else
-            unless LibCrypto.x509_verify_param_set1_host(param, hostname, 0) == 1
-              raise OpenSSL::Error.new("X509_VERIFY_PARAM_set1_host")
+      begin
+        if hostname
+          # Macro from OpenSSL: SSL_ctrl(s,SSL_CTRL_SET_TLSEXT_HOSTNAME,TLSEXT_NAMETYPE_host_name,(char *)name)
+          LibSSL.ssl_ctrl(
+            @ssl,
+            LibSSL::SSLCtrl::SET_TLSEXT_HOSTNAME,
+            LibSSL::TLSExt::NAMETYPE_host_name,
+            hostname.to_unsafe.as(Pointer(Void))
+          )
+
+          {% if compare_versions(LibSSL::OPENSSL_VERSION, "1.0.2") >= 0 %}
+            param = LibSSL.ssl_get0_param(@ssl)
+
+            if ::Socket.ip?(hostname)
+              unless LibCrypto.x509_verify_param_set1_ip_asc(param, hostname) == 1
+                raise OpenSSL::Error.new("X509_VERIFY_PARAM_set1_ip_asc")
+              end
+            else
+              unless LibCrypto.x509_verify_param_set1_host(param, hostname, 0) == 1
+                raise OpenSSL::Error.new("X509_VERIFY_PARAM_set1_host")
+              end
             end
-          end
-        {% else %}
-          context.set_cert_verify_callback(hostname)
-        {% end %}
-      end
+          {% else %}
+            context.set_cert_verify_callback(hostname)
+          {% end %}
+        end
 
-      ret = LibSSL.ssl_connect(@ssl)
-      unless ret == 1
-        raise OpenSSL::SSL::Error.new(@ssl, ret, "SSL_connect")
+        ret = LibSSL.ssl_connect(@ssl)
+        unless ret == 1
+          raise OpenSSL::SSL::Error.new(@ssl, ret, "SSL_connect")
+        end
+      rescue ex
```

Comment:
The `Server#initialize` looks good. But on `Client#initialize` still has the raise, rescue, raise.

Reply:
Client#initialize has far more return points..do you want me to insert it carefully before all of them?  It seems to me like it would clutter things quite a bit...

Reply:
It has come to my knowledge that even SSL_Accept can raise an Exception (ex "connection reset by peer") if the underlying read/writes within it fail.  Mind blown.  I propose we just go with begin rescue blocks?

Reply:
`ssl_accept` is a c function. How would it raise?

Reply:
`ssl_accept` makes calls to "read" from its BIO which calls back into Crystal land.  Crystal then calls "socket read" which can raise Crystal exceptions.  Surprised me...